dlib/dnn/input.h

```diff
@@ -690,11 +677,158 @@ namespace dlib
+
+    template <typename PYRAMID_TYPE>
+    class input_grayscale_image_pyramid : public input_image_pyramid<PYRAMID_TYPE>
+    {
+    public:
+        typedef matrix<unsigned char> input_type;
+        typedef PYRAMID_TYPE pyramid_type;
+
+        template <typename forward_iterator>
+        void to_tensor (
+                forward_iterator ibegin,
+                forward_iterator iend,
+                resizable_tensor& data
+        ) const
+        {
+            this->to_tensor_init(ibegin, iend, data, 1);
+
+            const auto rects = data.annotation().get<std::vector<rectangle>>();
+            if (rects.size() == 0)
+                return;
+
+            // copy the first raw image into the top part of the tiled pyramid.  We need to
+            // do this for each of the input images/samples in the tensor.
+            auto ptr = data.host_write_only();
+            for (auto i = ibegin; i != iend; ++i)
+            {
+                auto& img = *i;
+                ptr += rects[0].top()*data.nc();
+                for (long r = 0; r < img.nr(); ++r)
+                {
+                    auto p = ptr+rects[0].left();
+                    for (long c = 0; c < img.nc(); ++c)
+                        p[c] = (img(r,c))/256.0;
+                    ptr += data.nc();
+                }
+                ptr += data.nc()*(data.nr()-rects[0].bottom()-1);
+            }
+
+            this->create_tiled_pyramid(rects, data);
+        }
+
+        friend void serialize(const input_grayscale_image_pyramid& item, std::ostream& out)
+        {
+            serialize("input_grayscale_image_pyramid", out);
+            serialize(item.pyramid_padding, out);
+            serialize(item.pyramid_outer_padding, out);
+        }
+
+        friend void deserialize(input_grayscale_image_pyramid& item, std::istream& in)
+        {
+            std::string version;
+            deserialize(version, in);
+            if (version != "input_grayscale_image_pyramid")
+                throw serialization_error("Unexpected version found while deserializing dlib::input_grayscale_image_pyramid.");
+            deserialize(item.pyramid_padding, in);
+            deserialize(item.pyramid_outer_padding, in);
+        }
+
+        friend std::ostream& operator<<(std::ostream& out, const input_grayscale_image_pyramid& item)
+        {
+            out << "input_grayscale_image_pyramid()";
+            out << " pyramid_padding="<<item.pyramid_padding;
+            out << " pyramid_outer_padding="<<item.pyramid_outer_padding;
+            return out;
+        }
+
+        friend void to_xml(const input_grayscale_image_pyramid& item, std::ostream& out)
+        {
+            out << "<input_grayscale_image_pyramid"
+                <<"' pyramid_padding='"<<item.pyramid_padding
+                <<"' pyramid_outer_padding='"<<item.pyramid_outer_padding
+                <<"'/>";
+        }
+
+    private:
+        using input_image_pyramid<PYRAMID_TYPE>::pyramid_padding;
```

Comment:
Is this needed?  Do you get an error in the friend functions without this?

Reply:
I think it used to throw me a compilation error, but maybe the problem was somewhere else, and I thought it was that. I removed it and it's still working.